### DISCUSSION TITLE
12-bit JPEG/TIFF support

### DISCUSSION THREAD

**Discussion by tumagonx** [libvips/libvips#2001] [Ideas]
I can't open http://download.osgeo.org/gdal/data/gtiff/mandrilrgb_12.tif using nip2 or vipsdisp
I can view that image using built in Windows image viewer once I do:
tiffcp -c lzw mandrilrgb_12.tif mandrilrgb.tif

but vips refuse to open mandrilrgb_12.tif or mandrilrgb.tif

**Comment by tumagonx** [comment]
both tiffcp and vips using the same libtiff dll.

**Comment by jcupitt** [comment]
Hi, this is hard to fix, since libjpeg and libjpeg-turbo must be configured for either 8-bit or 12-bit, but not both. If you support 12-bit JPEG, you can't support 8-bit.

From the libjpeg install notes:

```
Support for 12-bit-deep pixel data:

The JPEG standard allows either 8-bit or 12-bit data precision.  (For color,
this means 8 or 12 bits per channel, of course.)  If you need to work with
deeper than 8-bit data, you can compile the IJG code for 12-bit operation.
To do so:
  1. In jmorecfg.h, define BITS_IN_JSAMPLE as 12 rather than 8.
  2. In jconfig.h, undefine BMP_SUPPORTED, RLE_SUPPORTED, and TARGA_SUPPORTED,
     because the code for those formats doesn't handle 12-bit data and won't
     even compile.  (The PPM code does work, as explained below.  The GIF
     code works too; it scales 8-bit GIF data to and from 12-bit depth
     automatically.)
  3. Compile.  Don't expect "make test" to pass, since the supplied test
     files are for 8-bit data.

Currently, 12-bit support does not work on 16-bit-int machines.

Note that a 12-bit version will not read 8-bit JPEG files, nor vice versa;
so you'll want to keep around a regular 8-bit compilation as well.
(Run-time selection of data depth, to allow a single copy that does both,
is possible but would probably slow things down considerably; it's very low
on our to-do list.)
```

**Comment by tumagonx** [comment]
I did the trick that gdal use, by having separate libjpeg linked to the same libtiff, so it could read both depth, I just wondering why vips still could not read it after turned to lzw compression? but that may be different question?

I'm agree regarding the performance note.

**Comment by jcupitt** [comment]
Sorry, I don't know. Could you link the 12-bit lzw version somewhere? I can't make it here. 

I tried with a regular jpg file and it seemed OK:

```
john@kiwi:~/pics$ vips copy k2.jpg k2.tif
john@kiwi:~/pics$ tiffcp -c lzw k2.tif k2-lzw.tif
john@kiwi:~/pics$ vipsheader k2-lzw.tif 
k2-lzw.tif: 1450x2048 uchar, 3 bands, srgb, tiffload
john@kiwi:~/pics$ vips copy k2-lzw.tif x.jpg
```

**Comment by tumagonx** [comment]
Try this link https://sourceforge.net/projects/pygobjectwin32/files/temp/mandrill.tif

I change it to zip compression for faster upload, it still readable by Windows but not by vips

**Comment by jcupitt** [comment]
It's a 12-bit packed file, so:

RR RG GG BB BR RR GG GB BB RR RG GG BB etc. etc

where each letter is four bits. libvips supports 1, 2, and 4 bit pallette RGB Images, but not unpalletted, unaligned multi-band. 

Let's turn this into an enhancement issue.

**Comment by tumagonx** [comment]
Ok, Thanks

**Comment by VioletGiraffe** [comment]
This isn't an RGB image, it's YCbCr. Just FYI. Windows Photos and ACDSee viewer cannot open it, neither can Photoshop CC 2017.
